activerecord/lib/active_record/relation/predicate_builder/polymorphic_array_handler.rb

```diff
@@ -0,0 +1,53 @@
+module ActiveRecord
+  class PredicateBuilder
+    class PolymorphicArrayHandler # :nodoc:
+      def initialize(predicate_builder)
+        @predicate_builder = predicate_builder
+      end
+
+      def call(attribute, value)
+        queries = {}
+
+        table = value.associated_table
+        queries = value.types_and_ids.map do |type, id|
+          { table.association_foreign_type.to_s => type, table.association_foreign_key.to_s => id }
+        end
+
+        predicates = queries.map { |query| predicate_builder.build_from_hash(query) }
+        type_and_ids_predicates = predicates.map { |type_predicate, id_predicate| Arel::Nodes::Grouping.new(type_predicate.and(id_predicate)) }
+        type_and_ids_predicates.inject { |composite, predicate| composite.or(predicate) }
```

Comment:
Is it worth grouping by type, so we can generate `x_type = 'A' AND x_id IN (1,2,3) OR x_type = 'B' AND x_id IN (4,5,6)`?

We generally avoid doing substantial contortions just for the sake of nicer SQL, but if it's relatively easy, the gain seems likely worthwhile here, IMO.